Restructure COALTON-LIBRARY/MATH

eliaslfox opened the pull request:
* move INTO and TRYINTO instances to library/math/conversions.lisp
* move NUM and other instances to library/math/num.lisp
* move FRACTION defintions to library/math/fraction.lisp

stylewarning commented on library/math/bounded.lisp:
```diff
@@ -0,0 +1,57 @@
+(in-package #:coalton-library/math/bounded)
+
+(coalton-toplevel
+  (define-class (Bounded :a)
```
How about IFIX and UFIX? You can use `(lisp Ifix () cl:most-positive-fixnum))` and the like.

stylewarning replied:
Also document the class.

stylewarning commented on library/math/conversions.lisp:
```diff
@@ -0,0 +1,566 @@
+  (define-instance (TryInto I32 I8)
+    (define (tryInto x)
+      (when (> x (into (the I8 maxBound)))
+        (return (Err "value out of range")))
```
any reason to not define a global constant

    (define out-of-range (Err ...))

stylewarning commented on library/math/conversions.lisp:
```diff
@@ -0,0 +1,566 @@
+      (when (> (into x) (the Integer (into (the I8 maxBound))))
+        (return (Err "value out of range")))
+
+      (Ok (lisp I8 (x) x))))
```
good ol `unsafe-reinterpret-cast`

stylewarning commented on library/math/conversions.lisp:
```diff
@@ -0,0 +1,566 @@
+  ;; Conversions from U8
+  ;;
+
+  (define-instance (TryInto U8 I8)
```
I think you'd have 4-5x fewer lines with

    (define-upconversion from-small to-big)

and

    (define-downconversion from-big to-small)

macros, or something.

eliaslfox replied:
Yeah probably. It's a little more annoying because all of the signed -> unsigned conversions aren't really upcasting or downcasting. Also I'd rather read the expanded code.

gefjon commented:
Should we just have `unsafe-reinterpret-cast` in the library, to simplify a common CL interop pattern and clarify intent?

aijony commented on library/math/conversions.lisp:
```diff
@@ -0,0 +1,566 @@
+      (lisp Integer (x) x))))
+
+(coalton-toplevel
+  (declare integer->double-float (Integer -> Double-Float))
```
This can result in an overflow for large integers that can't be represented by floats. Which is fine, but it should probably be documented, as well as perhaps a use-case to justify it. Otherwise, `fromInt` has achieves the same thing but with infs instead of errors. Although, `Into` instances of fixed-length integers would be completely safe.

eliaslfox replied:
Yeah we should just use fron int.

eliaslfox commented:
> Should we just have unsafe-reinterpret-cast in the library, to simplify a common CL interop pattern and clarify intent?

I don't think so. If we define `unsafe-case` as `:a -> :b` and it's used like:
```
(f (unsafe-cast x))
```
Changing the argument type of `f` to something else will still compile. Using casts with `lisp` requires specifying the intended output type.

eliaslfox commented:
* moved definitions for {IU..I64} from early types to `library/fixed-sized-numbers.lisp`
* added `Bounded` instances for `IFix` and `UFix`
* removed overflow checks from `IFix` as discussed
